File: internal/containers/distrobox.go
package containers

import (
	"context"
	"fmt"
	"strings"

	"github.com/iskry/cask/internal/config"
	"github.com/iskry/cask/internal/executor"
)

// DetectPackageManager returns the package manager name and install command
// based on the container image name.
func DetectPackageManager(image string) (string, []string) {
	lower := strings.ToLower(image)
	for _, d := range []string{"fedora", "centos", "rhel", "rocky", "alma"} {
		if strings.Contains(lower, d) {
			return "dnf", []string{"dnf", "install", "-y"}
		}
	}
	for _, d := range []string{"debian", "ubuntu", "mint"} {
		if strings.Contains(lower, d) {
			return "apt", []string{"apt-get", "install", "-y"}
		}
	}
	if strings.Contains(lower, "arch") {
		return "pacman", []string{"pacman", "-S", "--noconfirm"}
	}
	if strings.Contains(lower, "alpine") {
		return "apk", []string{"apk", "add"}
	}
	for _, d := range []string{"suse", "sles", "tumbleweed"} {
		if strings.Contains(lower, d) {
			return "zypper", []string{"zypper", "install", "-y"}
		}
	}
	return "dnf", []string{"dnf", "install", "-y"}
}

// BuildHomeFlags builds distrobox create flags from the home field.
func BuildHomeFlags(instance *config.DevboxInstance) []string {
	if instance.Home == "" || instance.Home == "host" {
		return nil
	}
	if instance.Home == "isolated" {
		return []string{
			"--no-entry", "--home",
			fmt.Sprintf("~/.local/share/distrobox/%s", instance.Name),
		}
	}
	return []string{"--home", instance.Home}
}

// SetupDevboxes creates Distrobox instances from config.
func SetupDevboxes(ctx context.Context, exec executor.Executor, cfg *config.DevboxConfig) error {
	if cfg == nil {
		return nil
	}

	for i := range cfg.Instances {
		inst := &cfg.Instances[i]

		cmd := []string{"distrobox", "create", "--name", inst.Name, "--image", inst.Image}
		cmd = append(cmd, BuildHomeFlags(inst)...)

		// Environment variables
		for key, value := range inst.Environment {
			cmd = append(cmd, "--additional-flags", fmt.Sprintf("--env %s=%s", key, value))
		}

		// Extra flags
		cmd = append(cmd, inst.Flags...)
		cmd = append(cmd, "--yes")

		r, err := exec.Execute(ctx, cmd)
		if err != nil {
			return fmt.Errorf("creating devbox %s: %w", inst.Name, err)
		}
		if !r.Success {
			return fmt.Errorf("creating devbox %s: %s", inst.Name, strings.TrimSpace(r.Stderr))
		}

		// Init hooks
		for _, hook := range inst.InitHooks {
			r, err = exec.Execute(ctx, []string{
				"distrobox", "enter", inst.Name, "--", "sh", "-c", hook,
			})
			if err != nil {
				return fmt.Errorf("init hook for %s: %w", inst.Name, err)
			}
		}

		// Install packages
		if len(inst.Packages) > 0 {
			_, installCmd := DetectPackageManager(inst.Image)
			pkgCmd := strings.Join(append(append([]string{"sudo"}, installCmd...), inst.Packages...), " ")
			exec.Execute(ctx, []string{
				"distrobox", "enter", inst.Name, "--", "sh", "-c", pkgCmd,
			})
		}

		// Post-create commands
		for _, postCmd := range inst.PostCreate {
			exec.Execute(ctx, []string{
				"distrobox", "enter", inst.Name, "--", "sh", "-c", postCmd,
			})
		}

		// Export apps
		for _, app := range inst.ExportApps {
			exec.Execute(ctx, []string{
				"distrobox", "enter", inst.Name, "--", "distrobox-export", "--app", app,
			})
		}
	}
	return nil
}

// ListDevboxes lists Distrobox instances.
func ListDevboxes(ctx context.Context, exec executor.Executor) ([]map[string]string, error) {
	r, err := exec.Execute(ctx, []string{"distrobox", "list", "--no-color"})
	if err != nil {
		return nil, err
	}
	if !r.Success {
		return nil, nil
	}
	var instances []map[string]string
	lines := strings.Split(r.Stdout, "\n")
	for _, line := range lines[1:] { // Skip header
		parts := strings.Split(line, "|")
		if len(parts) >= 3 {
			instances = append(instances, map[string]string{
				"id":     strings.TrimSpace(parts[0]),
				"name":   strings.TrimSpace(parts[1]),
				"status": strings.TrimSpace(parts[2]),
			})
		}
	}
	return instances, nil
}

File: src/cask/managers/distrobox.py
"""Distrobox instance management."""
from __future__ import annotations

from cask.config.models import DevboxInstanceConfig
from cask.executor.protocol import Executor
from cask.result import Result


def _detect_pkg_manager(image: str) -> tuple[str, list[str]]:
    """Return (pkg_manager_name, install_cmd_prefix) based on image."""
    image_lower = image.lower()
    if any(d in image_lower for d in ("fedora", "centos", "rhel", "rocky", "alma")):
        return "dnf", ["sudo", "dnf", "install", "-y"]
    elif any(d in image_lower for d in ("ubuntu", "debian", "mint")):
        return "apt", ["sudo", "apt-get", "install", "-y"]
    elif any(d in image_lower for d in ("arch", "manjaro", "endeavour")):
        return "pacman", ["sudo", "pacman", "-S", "--noconfirm"]
    elif any(d in image_lower for d in ("opensuse", "suse", "tumbleweed")):
        return "zypper", ["sudo", "zypper", "install", "-y"]
    elif any(d in image_lower for d in ("alpine",)):
        return "apk", ["sudo", "apk", "add"]
    else:
        return "dnf", ["sudo", "dnf", "install", "-y"]  # fallback


class DistroboxManager:
    """Manages Distrobox instances."""

    async def create(self, name: str, cfg: DevboxInstanceConfig, exec: Executor) -> Result:
        args = ["distrobox", "create", "--name", name, "--image", cfg.image, "--yes"]
        if cfg.home == "isolated":
            args.extend(["--home", f"~/.local/share/distrobox/{name}"])
        r = await exec.execute(args)
        if r.exit_code != 0:
            return Result(ok=False, message=f"Failed to create {name}: {r.stderr}", actions=[])

        actions = [f"Created distrobox {name}"]

        # Install packages if specified
        if cfg.packages:
            _, install_prefix = _detect_pkg_manager(cfg.image)
            pkg_cmd = f"distrobox enter {name} -- {' '.join(install_prefix)} {' '.join(cfg.packages)}"
            await exec.execute_shell(pkg_cmd)
            actions.append(f"Installed {len(cfg.packages)} packages in {name}")

        # Export apps
        for app in cfg.export_apps:
            await exec.execute(["distrobox", "enter", name, "--", "distrobox-export", "--app", app])
            actions.append(f"Exported {app}")

        return Result(ok=True, message=f"Created distrobox {name}", actions=actions)

    async def remove(self, name: str, exec: Executor) -> Result:
        r = await exec.execute(["distrobox", "rm", "--force", name])
        if r.exit_code == 0:
            return Result(ok=True, message=f"Removed distrobox {name}", actions=[])
        return Result(ok=False, message=f"Failed to remove {name}: {r.stderr}", actions=[])

    async def list_instances(self, exec: Executor) -> dict[str, str]:
        """Return {name: image} of distrobox instances."""
        r = await exec.execute(["distrobox", "list", "--no-color"])
        instances = {}
        if r.exit_code == 0:
            for line in r.stdout.strip().splitlines()[1:]:  # Skip header
                parts = [p.strip() for p in line.split("|")]
                if len(parts) >= 4:
                    instances[parts[1]] = parts[3]
        return instances
